reference/reference.go

```diff
@@ -0,0 +1,280 @@
+// Package reference provides a general type to represent any way of referencing images within the registry.
+// Its main purpose is to abstract tags and digests (content-addressable hash).
+//
+// Grammar
+//
+//     reference                       := repository [ ":" tag ] [ "@" digest ]
+//     repository                      := component [ [ port ] ( '/' component )* ]
+//     component                       := alpha-numeric [ separator alpha-numeric ]*
+//     alpha-numeric                   := /[a-z0-9]+/
+//     separator                       := /[._-]/
+//     port                            := /:[0-9]+/
+//     tag                             := /[\w][\w.-]{0,127}/
+//
+//     // from the digest package
+//     digest                          := digest-algorithm ":" digest-hex
+//     digest-algorithm                := digest-algorithm-component [ digest-algorithm-separator digest-algorithm-component ]
+//     digest-algorithm-separator      := /[+.-_]/
+//     digest-algorithm-component      := /[A-Za-z]/ /[A-Za-z0-9]*/
+//     digest-hex                      := /[A-Za-z0-9_-]+/ ; supports hex bytes or url safe base64
+package reference
+
+import (
+	"errors"
+	"regexp"
+
+	"github.com/docker/distribution/digest"
+)
+
+var (
+	tagRegexp    = regexp.MustCompile(`[\w][\w.-]{0,127}`)
+	alphaNumeric = regexp.MustCompile(`[a-z0-9]+`)
```

Comment:
I used the grammar you gave me, and this includes both hostname and repo name. Should we split that?